src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+                symbol_addr_table + i,                                                                  \
+                section->size,                                                                          \
+                restore_original);                                                                      \
+        continue;                                                                                       \
+    }
+        MEMRAY_HOOKED_FUNCTIONS
+#undef FOR_EACH_HOOKED_FUNCTION
+    }
+}
+
+static void
+patch_symbols_in_shared_object(const struct mach_header* header, intptr_t slide, bool restore_original)
+{
+    Dl_info info;
+    if (dladdr(header, &info) == 0) {
+        return;
+    }
+
+    const char* image_name = info.dli_fname;
+    if (strstr(image_name, "memray.cpython") || strstr(image_name, "/dyld")
+        || strstr(image_name, "dyld.dylib"))
+    {
+        LOG_IF_DEBUG("Skipping patching symbols for " << image_name);
+        return;
+    }
+    LOG_IF_DEBUG("Patching symbols for " << image_name);
+
+    auto segment_filter = [](const char* seg_name) {
+        return strcmp(seg_name, SEG_DATA) == 0 || strcmp(seg_name, SEG_DATA_CONST) == 0
+               || strcmp(seg_name, SEG_AUTH_CONST) == 0;
+    };
+
+    DynamicInfoTable dyninfo_table(header, slide, segment_filter);
+    if (!dyninfo_table) {
+        LOG_IF_DEBUG("Could not construct dynamic information table" << image_name);
+        return;
+    }
+
+    LOG_IF_DEBUG("Found " << dyninfo_table.segments.size() << " data segments");
+
+    for (const auto& segment_cmd : dyninfo_table.segments) {
+        const auto current_seg = reinterpret_cast<uintptr_t>(segment_cmd);
+        auto section_head = reinterpret_cast<const section_t*>(current_seg + sizeof(segment_command_t));
+        LOG_IF_DEBUG("Considering segment " << segment_cmd->segname);
+        for (size_t i = 0; i < segment_cmd->nsects; i++) {
+            const section_t* section = section_head + i;
+            unsigned int section_type = section->flags & SECTION_TYPE;
+            if (section_type != S_LAZY_SYMBOL_POINTERS && section_type != S_NON_LAZY_SYMBOL_POINTERS) {
+                LOG_IF_DEBUG("Skipping section" << i << " (" << section->segname << ")");
+                continue;
+            }
+            LOG_IF_DEBUG("Patching section number " << i << " (" << section->segname << ")");
+            patch_symbols_in_section(section, slide, dyninfo_table, restore_original);
+        }
+    }
+}
+
+static void
+_patch_symbols_in_shared_objects(const struct mach_header* header, intptr_t slide)
+{
+    patch_symbols_in_shared_object(header, slide, false);
+}
+
+static void
+_restore_symbols_in_shared_objects(const struct mach_header* header, intptr_t slide)
+{
+    patch_symbols_in_shared_object(header, slide, true);
+}
+
+/* Public API functions */
+
+void
+SymbolPatcher::overwrite_symbols() noexcept
+{
+    _dyld_register_func_for_add_image(_patch_symbols_in_shared_objects);
```

Comment:
This seems a bit weird - there's a register, but no unregister. Is it possible to register more than one function, or does calling register a second time unregister the first thing and register the second?

Reply:
Apparently it adds it a some kind of list:

https://github.com/apple-opensource/dyld/blob/1128192c016372ae94793d88530bc5978c1fce93/dyld3/AllImages.cpp#L1321-L1339

Reply:
We switched to iterating over the loaded libraries.